### PR TITLE
Add preliminary ALPN support for OpenSSL backend

### PR DESCRIPTION
**Pull request by seanmonstar**

### CODE REVIEW

**Comment by seanmonstar** [comment]
Seems CI failures are unhappy about lazy_static 0.2.

**Comment by sfackler** [comment]
I don't really want to support a feature that only works on one backend - the whole point of this crate is that it'll work anywhere!

File: src/imp/openssl.rs [hunk 1]
```diff
@@ -266,6 +266,16 @@ impl TlsConnector {
             connector.cert_store_mut().add_cert((cert.0).0.clone())?;
         }
 
+        let alpn = builder
+            .alpn_protocols
+            .iter()
+            .map(|proto| format!("\\x{:x}{}", proto.len(), proto))
```
Comment (sfackler) [inline src/imp/openssl.rs]: The sequence that openssl wants is literally the byte length of the protocol followed by it, not the string `\xN`.

**Comment by seanmonstar** [comment]
I assumed as mentioned in #49, support for backends could be added in as they exist.

**Comment by sfackler** [comment]
All of the underlying system APIs support ALPN.

File: src/imp/openssl.rs (hunk 1 above)
Comment (seanmonstar) [reply src/imp/openssl.rs]: Ah, my eyes tricked me when reading https://github.com/sfackler/hyper-openssl/blob/master/src/lib.rs#L107.

**Comment by seanmonstar** [comment]
I don't see anything about ALPN in the schannel docs...

**Comment by sfackler** [comment]
https://github.com/steffengy/schannel-rs/pull/49

**Comment by seanmonstar** [comment]
So to clarify, you'd rather not provide the API in `native-tls` until all 3 backends have the features available? I figured we could add support individually, since it's mostly an optimization, and things should still work if not available.

**Comment by sfackler** [comment]
> So to clarify, you'd rather not provide the API in `native-tls` until all 3 backends have the features available?

Yep.

**Pull request closed without merging**